databricks/koalas/frame.py

```diff
@@ -2977,6 +2977,86 @@ class  locomotion
             ).resolved_copy
             return DataFrame(internal)
 
+    def between_time(
+        self, start_time, end_time, include_start=True, include_end=True, axis: Union[int, str] = 0,
+    ) -> "DataFrame":
+        """
+        Select values between particular times of the day (e.g., 9:00-9:30 AM).
+
+        By setting ``start_time`` to be later than ``end_time``,
+        you can get the times that are *not* between the two times.
+
+        Parameters
+        ----------
+        start_time : datetime.time or str
+            Initial time as a time filter limit.
+        end_time : datetime.time or str
+            End time as a time filter limit.
+        include_start : bool, default True
+            Whether the start time needs to be included in the result.
+        include_end : bool, default True
+            Whether the end time needs to be included in the result.
+        axis : {0 or 'index', 1 or 'columns'}, default 0
+            Determine range time on index or columns value.
+
+        Returns
+        -------
+        Series or DataFrame
+            Data from the original object filtered to the specified dates range.
+
+        Raises
+        ------
+        TypeError
+            If the index is not  a :class:`DatetimeIndex`
+
+        See Also
+        --------
+        at_time : Select values at a particular time of the day.
+        first : Select initial periods of time series based on a date offset.
+        last : Select final periods of time series based on a date offset.
+        DatetimeIndex.indexer_between_time : Get just the index locations for
+            values between particular times of the day.
+
+        Examples
+        --------
+        >>> i = pd.date_range('2018-04-09', periods=4, freq='1D20min')
+        >>> ts = pd.DataFrame({'A': [1, 2, 3, 4]}, index=i)
+        >>> kts = ks.from_pandas(ts)
+        >>> kts
+                             A
+        2018-04-09 00:00:00  1
+        2018-04-10 00:20:00  2
+        2018-04-11 00:40:00  3
+        2018-04-12 01:00:00  4
+
+        >>> kts.between_time('0:15', '0:45')
+                             A
+        2018-04-10 00:20:00  2
+        2018-04-11 00:40:00  3
+
+        You get the times that are *not* between two times by setting
+        ``start_time`` later than ``end_time``:
+
+        >>> kts.between_time('0:45', '0:15')
+                             A
+        2018-04-09 00:00:00  1
+        2018-04-12 01:00:00  4
+        """
+        from databricks.koalas.indexes import DatetimeIndex
+
+        if axis == 0:
+            index = self.index
+        else:
+            raise NotImplementedError("between_time currently only works for axis=0")
+
+        if not isinstance(index, DatetimeIndex):
+            raise TypeError("Index must be DatetimeIndex")
+
+        indexer = index.indexer_between_time(
+            start_time, end_time, include_start=include_start, include_end=include_end
+        ).to_numpy()
```

Comment:
I am not sure `to_numpy()` is a good idea since all the data will be loaded into the driver's memory.
Let me think about this.

Reply:
The reason I used `to_numpy()` is because without it I got this not implemented error.

`databricks.koalas.exceptions.PandasNotImplementedError: The method `pd.Index.__iter__()` is not implemented. If you want to collect your data as an NumPy array, use 'to_numpy()' instead.`

Reply:
How about
```
       ...
        def pandas_between_time(pdf):
            return pdf.between_time(start_time, end_time, include_start, include_end)

        return self.koalas.apply_batch(pandas_between_time)
```
?

Reply:
We might also want to add a test
```
    def test_between_time_no_shortcut(self):
        with ks.option_context("compute.shortcut_limit", 0):
            i = pd.date_range("2018-04-09", periods=4, freq="1D20min")
            ts = pd.DataFrame({"A": [1, 2, 3, 4]}, index=i)
            kts = ks.DataFrame({"A": [1, 2, 3, 4]}, index=i)
            self.assert_eq(
                ts.between_time("0:15", "0:45"), kts.between_time("0:15", "0:45"), almost=True
            )
```